pkg/apis/sparkscheduler/v1beta1/conversion_resource_reservation.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2019 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1beta1
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/apis/sparkscheduler/v1beta2"
+	werror "github.com/palantir/witchcraft-go-error"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+)
+
+// ConvertTo converts from v1beta1 to the storage version v1beta2
+// We first try to take values the v1beta1 struct, we then take all remaining values from the annotations.
+func (rr *ResourceReservation) ConvertTo(dstRaw conversion.Hub) error {
+	dst, ok := dstRaw.(*v1beta2.ResourceReservation)
+	if !ok {
+		return werror.Error("dst type not as expected",
+			werror.SafeParam("expectedType", fmt.Sprintf("%T", v1beta2.ResourceReservation{})),
+			werror.SafeParam("actualType", fmt.Sprintf("%T", dstRaw)))
+	}
+
+	dst.ObjectMeta = *rr.ObjectMeta.DeepCopy()
+
+	// Remove the reservation annotation metadata as we don't need it in a v2 object.
+	delete(dst.ObjectMeta.Annotations, ReservationsAnnotation)
+
+	dst.Status.Pods = make(map[string]string, len(rr.Status.Pods))
+	for key, value := range rr.Status.Pods {
+		dst.Status.Pods[key] = value
+	}
+
+	// Take the common values from the v1 struct
+	dst.Spec.Reservations = make(map[string]v1beta2.Reservation, len(rr.Spec.Reservations))
+	for key, value := range rr.Spec.Reservations {
+		dst.Spec.Reservations[key] = v1beta2.Reservation{
+			Node: value.Node,
+			Resources: v1beta2.ResourceList{
+				string(v1beta2.ResourceCPU):    &value.CPU,
+				string(v1beta2.ResourceMemory): &value.Memory,
+			},
+		}
+	}
+
+	// Attempt to take any other values from the ReservationsAnnotation
+	if annotationResourceReservationSpecJSON, ok := rr.ObjectMeta.Annotations[ReservationsAnnotation]; ok {
+		var annotationResourceReservationSpec v1beta2.ResourceReservationSpec
+		err := json.Unmarshal([]byte(annotationResourceReservationSpecJSON), &annotationResourceReservationSpec)
+		if err != nil {
+			return err
+		}
+		for key, annotationReservation := range annotationResourceReservationSpec.Reservations {
+			// If the reservation did not exist in the reservations of the v1 struct,
+			// make the reservation with an empty resource list
+			if val, ok := dst.Spec.Reservations[key]; !ok {
+				dst.Spec.Reservations[key] = v1beta2.Reservation{
+					Node:      val.Node,
+					Resources: v1beta2.ResourceList{},
+				}
```

Comment:
When do we expect this scenario to occur? Given `v1beta2` is the storage version, the only (natural) time you'll have a `v1beta1` with the annotation is if it passed through the webhook which should have generated the `v1beta1` object with the right reservations, no?

Reply:
Note: a reservation could have been removed by the user of `v1beta1`, but in this instance I think we don't want to reintroduce it